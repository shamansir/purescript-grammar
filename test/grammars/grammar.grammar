main :- repSep((ruleDefn | comment), ws).

comment :- ["#", repSep(commentChar, "")].
ruleDefn :- [ident, ws, ":-", ws, rule, "."].

rule :- (seq | choice | ref | text | charRule | repSep | placeholder).

seq :- ["[", ws, repSep(rule, [ws, ",", ws]), ws, "]"].
choice :- ["(", ws, repSep(rule, [ws, "|", ws]), ws, ")"].
ref :- [([captureName, ":"] | ""), ruleName].
captureName :- ident.
ruleName :- ident.
text :- ["\"", repSep(stringChar, ""), "\""].
repSep :- [repSepKW, "(", rep:rule, commaSpace, sep:rule, ")"].
repSepKW :- "repSep".

charRule :- (charRange | notChar | singleChar | anyChar).
charRange :- ["[", from:alphaNum, "-", to:alphaNum, "]"].
notChar :- ['^', charRule].
singleChar :- ["'", (['\\', 'n'] | ['\\', '\\'] | .), "'"].
anyChar :- ".".

ident :- repSep(alpha, "").
ws :- repSep((" "|"\n"), "").
commaSpace :- [",", ws].
placeholder :- "???".

alphaNum :- (alpha | num).
alpha :- ([a-z] | [A-Z]).
num :- [0-9].
stringChar :- (^'"' | ['\\', '"']).
commentChar :- ^'\n'.
