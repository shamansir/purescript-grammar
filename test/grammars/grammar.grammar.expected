main :- repSep((ruleDefn|comment),ws).
alpha :- ([a-z]|[A-Z]).
alphaNum :- (alpha|num).
anyChar :- ".".
captureName :- ident.
charRange :- ["[",from:alphaNum,"-",to:alphaNum,"]"].
charRule :- (charRange|notChar|singleChar|anyChar).
choice :- ["(",ws,repSep(rule,[ws,"|",ws]),ws,")"].
commaSpace :- [",",ws].
comment :- ["#",repSep(commentChar,"")].
commentChar :- ^'\n'.
ident :- repSep(alpha,"").
notChar :- ['^',charRule].
num :- [0-9].
placeholder :- "???".
ref :- [([captureName,":"]|""),ruleName].
repSep :- [repSepKW,"(",rep:rule,commaSpace,sep:rule,")"].
repSepKW :- "repSep".
rule :- (seq|choice|ref|text|charRule|repSep|placeholder).
ruleDefn :- [ident,ws,":-",ws,rule,"."].
ruleName :- ident.
seq :- ["[",ws,repSep(rule,[ws,",",ws]),ws,"]"].
singleChar :- ["'",(['\\','n']|['\\','\\']|.),"'"].
stringChar :- (^'"'|['\\','"']).
text :- ["\"",repSep(stringChar,""),"\""].
ws :- repSep((" "|"\n"),"").